Marlin/src/module/tool_change.cpp

```diff
@@ -384,12 +388,88 @@ inline void fast_line_to_current(const AxisEnum fr_axis) { _line_to_current(fr_a
 
 #if ENABLED(SWITCHING_TOOLHEAD)
 
+  // Return a bitmask of tool sensor states
+  static inline uint8_t poll_tool_sensor_pins() {
+    return (0
+      #if ENABLED(TOOL_SENSOR)
+        #if PIN_EXISTS(TOOL_SENSOR1)
+          | (READ(TOOL_SENSOR1_PIN) ? _BV(0) : 0)
+        #endif
+        #if PIN_EXISTS(TOOL_SENSOR2)
+          | (READ(TOOL_SENSOR2_PIN) ? _BV(1) : 0)
+        #endif
+        #if PIN_EXISTS(TOOL_SENSOR3)
+          | (READ(TOOL_SENSOR3_PIN) ? _BV(2) : 0)
+        #endif
+      #endif
+    );
+  }
+
+  #if ENABLED(TOOL_SENSOR)
+
+    bool disable_sensor; // = false
+
+    uint8_t check_tool_sensor_stats(uint8_t active_tool, bool kill_on_error, bool disable) {
+      static uint8_t sensor_try_Again; // = 0
+      for (;;) {
+        if (poll_tool_sensor_pins() == _BV(active_tool)) {
+          sensor_try_Again = 0;
+          return active_tool;
+        }
+        else if (kill_on_error == true && (disable_sensor == false || disable == true)) {
+          sensor_try_Again++;
+          if (sensor_try_Again > 10) {
+            std::string status = "TS error " + std::to_string(active_tool);
+            kill(status.c_str());
+          }
+          safe_delay(5);
+        }
+        else {
+          sensor_try_Again++;
+          if (sensor_try_Again > 10) return -1;
+          safe_delay(5);
+        }
+      }
+    }
+
+  #else
+    inline uint8_t check_tool_sensor_stats(uint8_t, bool, bool) {}
+  #endif
+
   inline void swt_lock(const bool locked=true) {
-    const uint16_t swt_angles[2] = SWITCHING_TOOLHEAD_SERVO_ANGLES;
-    MOVE_SERVO(SWITCHING_TOOLHEAD_SERVO_NR, swt_angles[locked ? 0 : 1]);
+    //const uint16_t swt_angles[2] = SWITCHING_TOOLHEAD_SERVO_ANGLES;
+    //MOVE_SERVO(SWITCHING_TOOLHEAD_SERVO_NR, swt_angles[locked ? 0 : 1]);
+    OUT_WRITE(SOL0_PIN, locked);
+    gcode.dwell(10);
```

Comment:
This should be altered based on some new option, which might be called `SWITCHING_TOOLHEAD_SOLENOID`. By default a switching toolhead employs a servo.

Reply:
Your completely right, I was using this as a workaround for servo issues and was just meant as temporary. (I have an external arduino that changes solenoid output to pwm for the servo) 
Most of the code was wip I just didn't have time yet to look at it. 
Thanks for going trough the code and making so many great fixes :)